invenio_sipstore/tasks.py

```diff
@@ -0,0 +1,58 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2016 CERN.
+#
+# Invenio is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Invenio is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Background tasks for sipstore."""
+
+from __future__ import absolute_import, print_function
+
+import os
+import shutil
+import tempfile
+
+from bagit import make_bag
+from celery import shared_task
+
+
+@shared_task
+def generate_bagit_file(record_sip):
+    """Celery task to create a bagit file from a sip package."""
+    path = tempfile.mkdtemp()
+
+    files = record_sip.sip_files
+    files_to_fs(files, path)
+
+    result = make_bag(path, {'Contact-Name': 'Test'})
+    result.save()
+
+    shutil.rmtree(path)
+    return result
+
+
+def files_to_fs(files, path):
+    """Copy the files list to the path specified."""
+    for file in files:
+        filename = os.path.join(path, file.filepath)
+        op = open(filename, mode='w')
+        ip = file.file.storage().open()
+        op.write(ip.read())
```

Comment:
@lnielsen This method is copying the files to a temporary directory in order to make use of the bagit library. Here is the inefficient point as you were commenting.